lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Factory.h"
+#include "cudaq/Optimizer/CodeGen/Passes.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_UNITARYSYNTHESIS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "unitary-synthesis"
+
+using namespace mlir;
+
+namespace cudaq {
+std::vector<std::complex<double>>
+readGlobalConstantArray(cudaq::cc::GlobalOp &global);
+}
+
+namespace {
+
+struct EulerAngles {
+  double alpha;
+  double beta;
+  double gamma;
+};
+
+struct BasisZYZ {
+
+  std::array<std::complex<double>, 4> matrix;
+  EulerAngles angles;
+  /// Updates to the global phase
+  double phase;
+
+  /// This logic is based on https://arxiv.org/pdf/quant-ph/9503016 and its
+  /// corresponding explanation in https://threeplusone.com/pubs/on_gates.pdf,
+  /// Section 4.
+  void decompose() {
+    using namespace std::complex_literals;
+
+    /// Rescale the input unitary matrix, `u`, to be special unitary.
+    /// Extract a phase factor, `phase`, so that `det(inv_phase * unitary) = 1`
+    auto det = (matrix[0] * matrix[3]) - (matrix[1] * matrix[2]);
+    phase = 0.5 * std::atan2(det.imag(), det.real());
+    auto inv_phase = std::exp(-1i * phase);
+
+    std::array<std::complex<double>, 4> special_unitary;
+    std::transform(matrix.begin(), matrix.end(), special_unitary.begin(),
+                   [&](auto element) { return element * inv_phase; });
+
+    auto abs_00 = std::abs(special_unitary[0]);
+    auto abs_01 = std::abs(special_unitary[1]);
```

Comment:
nit: No snake_case, Thanks. There is more snake_case here in this function, but I will stop with the nits here.

Reply:
Fixed. 3e71dee865f261ba1ba9379e172eb4b68ccab2d7